### PR TITLE
IR: add support for `switch` statements

### LINKED ISSUE #273: Support `switch` statements in IR
The basic feature shouldn't be too hard: just introduce a new kind of terminator instruction that takes (3 + 2*N) operands:

    switch <integerVal> <breakLabel> <defaultLabel> [<firstCaseVal> <firstCaseLabel>]*

The `breakLabel` should be the block after the `switch` ends. The `defaultLabel` should be the `default:` case if there is one, and the `breakLabel` otherwise. The `firstCaseVal`, etc. should all be integer literal values.

For the purposes of our current source emit strategy, we should probably assume/require that the `case`s are represented in source order.

Where this will get a bit more complicated is when we get around to introducing proper SSA into the IR, so long as we stick with the current "blocks with parameters" model (which I'd like to try). The catch here is that we are already using the "varargs" capability of `IRInst` to store the N cases, so being able to store arguments for the per-case blocks would just complicate things.

A simple solution is to declare that a `switch` instruction can only branch to blocks with zero parameters. This shouldn't ever be a problem in practice because:

- A `case` will almost always have only a single predecessor (the block with the `switch`) so there'd be no reason for it to take parameters (equivalently: there is no reason why it would contain `phi` instructions).

- In the rare circumstance that a `case` block needs arguments (this could occur with non-trivial use of fall-through) we could insert additional blocks to break the critical edge, and have that new block do the argument-passing.

### PR DESCRIPTION
**Pull request by tfoleyNV**
Fixes #273

This is just something we hadn't gotten to yet on the IR. The actual design of the instruction is unsurprising (once you take into consideration the requirement for structured control flow). A `switch` instruction takes the form:

    switch <condition> <breakLabel> <defaultLabel> [<caseVal> <caseLabel>]*

Where `condition` is the value to switch on, `breakLabel` is the "join point" after the original `switch` statement, `defaultLabel` is where to go if the value doesn't match any case, and each pair of `caseVal` and `caseLabel` is what to do on a particular value. It is required that `caseVal` be a literal, but this isn't currently being enforced in the IR (the front-end should be making a check and constant-folding the case labels).

For structured control flow, we also make the assumption that the cases are in order: cases with the same label must be grouped together, and any case that falls through to another must come right before it.

Given this representation, the emit logic can reconstruct a `switch` statement with relative ease, given the machinery we already have. It makes sure to group together case values with the same label (again, assuming they are contiguous), and will insert the `default:` label in with whatever group it belongs to.

Actually emitting code for a `switch` statement seems superficially simple, until you realize that a complete implementation needs to handle stuff like "Duff's Device." The current implementation makes the assumption that all `case` and `default` statements are directly nested under a `switch`, and that there is no way for control flow to enter a case except by the `switch` itself, or fall-through.

In order to facilitate the grouping of cases in the IR-to-HLSL emit logic, the AST-to-IR lowering logic tries to detect cases where there are multiple `case`s in a row, and emit only a single label for them.

One big/annoying gotcha is that we don't properly handle the case where a `default:` case has a non-trivial fall-throguh to another case. That seems fine for now since HLSL doesn't support fall-through anyway, but it probably needs to get detected somewhere in the Slang compiler (e.g., maybe we should add a diagnostic pass over the IR that detects target-specific problems like that and emits errors).

### CODE REVIEW

**Comment by tfoleyNV** [comment]
@csyonghe Wait a second on the review. I'm going to add another fix to this PR so that we handle `EmptyStmt` as well (seems to minor to get its own PR).

**Comment by tfoleyNV** [comment]
@csyonghe Okay, this should be ready to review now.

File: tests/compute/switch-stmt.slang.1.expected.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+10
```
Comment (csyonghe) [inline tests/compute/switch-stmt.slang.1.expected.txt]: What is this file for?

**Comment by tfoleyNV** [comment]
@csyonghe On the subject of `break`s and nesting: the front-end keeps track of the "parent" statement for any `break` or `continue`, and during lowering to IR they will target the label that has been registered for their "parent" statement. This means that the front end and IR lowering can technically handle multi-level (or named) `break` if we ever wanted to add it.

For the IR->HLSL stuff, we basically always turn a `break` instruction into a `break` statement, and assume that the IR doesn't include anything that encodes a multi-level break.

So it all *should* Just Work for now, but if we ever introduce multi-level breaks (even if just as a result of doing inlining optimizations), we'll need to revise the IR->source pass.

File: tests/compute/switch-stmt.slang.1.expected.txt (hunk 1 above)
Comment (tfoleyNV) [reply tests/compute/switch-stmt.slang.1.expected.txt]: If a test file has multiple `//TEST:...` lines, then we will use the same file to run multiple tests, where the "output stem" will take the form `name`, then `name.1`, then `name.2`, etc.

So this file is the expected output for the second `//TEST:...` line in `switch-stmt.slang`.

**Pull request merged**